src/coprocessor/dag/expr/builtin_time.rs

```diff
@@ -24,9 +24,20 @@ impl ScalarFunc {
         ctx: &mut EvalContext,
         row: &'a [Datum],
     ) -> Result<Option<Cow<'a, [u8]>>> {
-        let t = try_opt!(self.children[0].eval_time(ctx, row));
+        let t = direct_try!(
```

Comment:
Personally I don't like `direct_try`. How about write it like this:
```rust
        let e = match self.children[0].eval_time(ctx, row)? {
            Some(res) => if !res.invalid_zero() {
                let format_mask = try_opt!(self.children[1].eval_string_and_decode(ctx, row));
                let res = res.date_format(format_mask.into_owned())?;
                return Ok(Some(Cow::Owned(res.into_bytes())));
            } else {
                Error::incorrect_datetime_value(&format!("Incorrect datetime value: '{}'", res))
            },
            None => Error::incorrect_datetime_value("None"),
        };
        Error::handle_invalid_time_error(ctx, e)?;
        Ok(None)
```
Please remember to use `eval_string_and_decode` if you want to decode the argument to string.

And, could you help us to let `date_format` retrieve an `&str` instead of `String`? You can do it at any convenient time.